pkg/operator/controllers/muo/deploy.go

```diff
@@ -0,0 +1,115 @@
+package muo
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	extensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
+)
+
+type Interface interface {
+	CreateOrUpdate(context.Context) error
+	Remove(context.Context) error
+	//IsReady(context.Context) (bool, error)
+}
+
+type muoDeployer struct {
+	log     *logrus.Entry
+	cluster *arov1alpha1.Cluster
+
+	extensionscli extensionsclient.Interface
+	dh            dynamichelper.Interface
+}
+
+func NewMUODeployer(log *logrus.Entry, cluster *arov1alpha1.Cluster, extensionscli extensionsclient.Interface, dh dynamichelper.Interface) Interface {
+	return &muoDeployer{
+		log:     log,
+		cluster: cluster,
+
+		extensionscli: extensionscli,
+		dh:            dh,
+	}
+}
+
+func (o *muoDeployer) resources() ([]kruntime.Object, error) {
+	results := []kruntime.Object{}
+	for _, assetName := range AssetNames() {
+		b, err := Asset(assetName)
+		if err != nil {
+			return nil, err
+		}
+
+		obj, _, err := scheme.Codecs.UniversalDeserializer().Decode(b, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// set the image for the deployments
+		if d, ok := obj.(*appsv1.Deployment); ok {
+			if d.Labels == nil {
+				d.Labels = map[string]string{}
+			}
+
+			imagePullspec, ext := o.cluster.Spec.OperatorFlags.Get(PULLSPEC)
+			if !ext {
+				return nil, errors.New("missing pullspec")
+			}
+
+			for i := range d.Spec.Template.Spec.Containers {
+				d.Spec.Template.Spec.Containers[i].Image = imagePullspec
+			}
+		}
+
+		results = append(results, obj)
+	}
+
+	return results, nil
+}
+
+func (o *muoDeployer) CreateOrUpdate(ctx context.Context) error {
+	resources, err := o.resources()
+	if err != nil {
+		return err
+	}
+
+	err = dynamichelper.Prepare(resources)
+	if err != nil {
+		return err
+	}
+
+	err = o.dh.Ensure(ctx, resources...)
+	return err
+}
+
+func (o *muoDeployer) Remove(ctx context.Context) error {
+	for _, assetName := range AssetNames() {
```

Comment:
Do we care about the order? `AssetNames()` returns a slice which is converted from a `map`. Iterating over a map in go is randomised.

Reply:
I think, it is not a problem in `CreateOrUpdate` becasue `dynamichelper.Prepare(resources)` deals with the order as far as I can see, but can it cause issues with deletion?

Reply:
If we want to delete the whole thing, yes, but right now we are only deleting a single deployment, so I didn't bother.